include/task_manager.h

```diff
@@ -47,42 +43,53 @@ namespace detail {
 		}
 
 	  private:
-		task_id m_this_epoch;
+		task_id m_this_epoch = 0;
 		mutable std::mutex m_mutex;
 		mutable std::condition_variable m_epoch_changed;
 	};
 
+	// definition is in handler.h to avoid circular dependency
+	template <typename CGF>
+	std::unique_ptr<task> invoke_command_group_function(const task_id tid, size_t num_collective_nodes, CGF&& cgf);
+
 	class task_manager {
 		friend struct task_manager_testspy;
 
 	  public:
+		/// Implement this as the owner of task_manager to receive callbacks on generated tasks.
+		class delegate {
+		  protected:
+			delegate() = default;
+			delegate(const delegate&) = default;
+			delegate(delegate&&) = default;
+			delegate& operator=(const delegate&) = default;
+			delegate& operator=(delegate&&) = default;
+			~delegate() = default; // do not allow destruction through base pointer
+
+		  public:
+			/// Called whenever new tasks have been generated and inserted into the tasks graph.
+			virtual void notify_task_created(const task* tsk) = 0; // TODO rename
+		};
+
 		struct policy_set {
 			error_policy uninitialized_read_error = error_policy::panic;
 		};
 
-		constexpr inline static task_id initial_epoch_task = 0;
-
-		task_manager(size_t num_collective_nodes, detail::task_recorder* recorder, const policy_set& policy = default_policy_set());
+		task_manager(size_t num_collective_nodes, detail::task_recorder* recorder, delegate* dlg, const policy_set& policy = default_policy_set());
 
 		virtual ~task_manager() = default;
 
 		template <typename CGF>
-		task_id submit_command_group(CGF cgf) {
-			auto reservation = m_task_buffer.reserve_task_entry(await_free_task_slot_callback());
-			const auto tid = reservation.get_tid();
-
-			handler cgh = make_command_group_handler(tid, m_num_collective_nodes);
-			cgf(cgh);
-
-			auto unique_tsk = into_task(std::move(cgh));
-
-			auto& tsk = register_task_internal(std::move(reservation), std::move(unique_tsk));
+		task_id submit_command_group(CGF&& cgf) {
+			const auto tid = m_next_tid++;
+			auto unique_tsk = invoke_command_group_function(tid, m_num_collective_nodes, std::forward<CGF>(cgf));
+			auto& tsk = register_task_internal(std::move(unique_tsk));
 			compute_dependencies(tsk);
 
 			// the following deletion is intentionally redundant with the one happening when waiting for free task slots
 			// we want to free tasks earlier than just when running out of slots,
 			// so that we can potentially reclaim additional resources such as buffers earlier
-			m_task_buffer.delete_up_to(m_latest_epoch_reached.get());
+			m_task_graph.prune_before_epoch(m_latest_epoch_reached.get());
```

Comment:
I went with `delete_before_epoch`, it's important that `epoch` remains a substring to recognize it as the counterpart to `begin_epoch` (since the data structure does not impose any restriction on the nodes inserted into one epoch)

Reply:
I like that.